src/ast/effects.h

```diff
@@ -48,18 +48,18 @@ struct EffectAnalyzer : public PostWalker<EffectAnalyzer> {
   bool readsMemory = false;
   bool writesMemory = false;
   bool implicitTrap = false; // a load or div/rem, which may trap. we ignore trap
-                             // differences, so it is ok to reorder these, and we
-                             // also allow reordering them with other effects
-                             // (so a trap may occur later or earlier, if it is
-                             // going to occur anyhow), but we can't remove them,
-                             // they count as side effects
+                             // differences, so it is ok to reorder these, but we can't
+                             // remove them, as they count as side effects, and we
+                             // can't move them in a way that would cause other noticeable
+                             // (global) side effects
   bool isAtomic = false; // An atomic load/store/RMW/Cmpxchg or an operator that
                          // has a defined ordering wrt atomics (e.g. grow_memory)
 
   bool accessesLocal() { return localsRead.size() + localsWritten.size() > 0; }
   bool accessesGlobal() { return globalsRead.size() + globalsWritten.size() > 0; }
   bool accessesMemory() { return calls || readsMemory || writesMemory; }
   bool hasSideEffects() { return calls || localsWritten.size() > 0 || writesMemory || branches || globalsWritten.size() > 0 || implicitTrap || isAtomic; }
+  bool hasGlobalSideEffects() { return calls || globalsWritten.size() > 0 || writesMemory; }
```

Comment:
Hmm, not sure I follow. Without adding atomic here, it's possible that we move a load across an atomic load. The load may trap, so if we move it before vs after the atomic load, the atomic synchronization may or may not happen. Are you saying that is something we should work to preserve? Or is a better example?